File: jetbrains-plugin/src/main/kotlin/com/vscode/jetbrainssync/OperationQueueProcessor.kt
package com.vscode.jetbrainssync

import com.intellij.openapi.diagnostic.Logger
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Operation queue processor
 * Responsible for handling asynchronous operation queues, ensuring atomicity and order of file operations
 * Includes queue capacity management and operation addition logic
 */
class OperationQueueProcessor(
    private val messageSender: MessageSender,
    private val localIdentifierManager: LocalIdentifierManager
) {
    private val log: Logger = Logger.getInstance(OperationQueueProcessor::class.java)

    // Internal queue management
    private val operationQueue = LinkedBlockingQueue<EditorState>()
    private val maxQueueSize = 100

    // Thread pool
    private val executorService: ExecutorService = Executors.newSingleThreadExecutor { r ->
        val thread = Thread(r, "Operation-Queue-Processor")
        thread.isDaemon = true
        thread
    }

    // Processing state
    private val isShutdown = AtomicBoolean(false)

    init {
        // Automatically start queue processor in constructor
        start()
    }

    /**
     * Add operation to queue
     * Includes queue capacity management logic
     */
    fun addOperation(state: EditorState) {
        if (operationQueue.size >= maxQueueSize) {
            operationQueue.poll()
            log.warn("Operation queue is full, removing oldest operation")
        }

        operationQueue.offer(state)
        log.info("Operation pushed to queue: ${state.action} ${state.filePath}, ${state.getCursorLog()}, ${state.getSelectionLog()}")
    }

    /**
     * Start queue processor
     */
    fun start() {
        log.info("Starting operation queue processor")
        executorService.submit(this::processQueue)
    }

    /**
     * Queue processing main loop
     */
    private fun processQueue() {
        while (!isShutdown.get() && !Thread.currentThread().isInterrupted) {
            try {
                // Block and wait for tasks in queue
                val state = operationQueue.take()
                processOperation(state)

                // Avoid overly frequent operations
                Thread.sleep(50)
            } catch (e: InterruptedException) {
                Thread.currentThread().interrupt()
                break
            } catch (e: Exception) {
                log.warn("Queue processor error: ${e.message}", e)
            }
        }
    }

    /**
     * Process single operation
     */
    private fun processOperation(state: EditorState) {
        try {
            sendStateUpdate(state)
        } catch (e: Exception) {
            log.warn("Failed to process operation: ${e.message}", e)
        }
    }

    /**
     * Send state update
     */
    private fun sendStateUpdate(state: EditorState) {
        val messageWrapper = MessageWrapper.create(
            localIdentifierManager.generateMessageId(),
            localIdentifierManager.identifier,
            state
        )
        val success = messageSender.sendMessage(messageWrapper)
        if (success) {
            log.info("✅ Message sent: ${state.action} ${state.filePath}, ${state.getCursorLog()}, ${state.getSelectionLog()}")
        } else {
            log.info("❌ Failed to send message: ${state.action} ${state.filePath}, ${state.getCursorLog()}, ${state.getSelectionLog()}")
        }
    }


    /**
     * Stop processor
     */
    fun dispose() {
        log.info("Starting to shutdown operation queue processor")

        isShutdown.set(true)
        executorService.shutdown()

        try {
            if (!executorService.awaitTermination(5, TimeUnit.SECONDS)) {
                executorService.shutdownNow()
            }
        } catch (e: InterruptedException) {
            executorService.shutdownNow()
        }

        // Clear queue
        operationQueue.clear()

        log.info("Operation queue processor shutdown completed")
    }
}

File: vscode-extension/src/OperationQueueProcessor.ts
import {EditorState, MessageWrapper, MessageSender} from './Type';
import {Logger} from './Logger';
import {LocalIdentifierManager} from './LocalIdentifierManager';

/**
 * Operation queue processor
 * Ensures atomicity and order of operations
 * Includes queue capacity management and operation addition logic
 */
export class OperationQueueProcessor {
    private logger: Logger;
    private messageSender: MessageSender;
    private processingInterval: NodeJS.Timeout | null = null;

    // Internal queue management
    private operationQueue: EditorState[] = [];
    private maxQueueSize = 100;

    // Processing state
    private isShutdown: boolean = false;

    constructor(
        logger: Logger,
        messageSender: MessageSender,
    ) {
        this.logger = logger;
        this.messageSender = messageSender;

        // Automatically start queue processor in constructor
        this.start();
    }

    /**
     * Add operation to queue
     * Includes queue capacity management logic
     */
    addOperation(state: EditorState) {
        if (this.operationQueue.length >= this.maxQueueSize) {
            this.operationQueue.shift();
            this.logger.warn('Operation queue is full, removing oldest operation');
        }

        this.operationQueue.push(state);
        this.logger.info(`Operation pushed to queue: ${state.action} ${state.filePath}, ${state.getCursorLog()}, ${state.getSelectionLog()}`)
    }

    /**
     * Start queue processor
     */
    start() {
        if (!this.isShutdown) {
            this.logger.info('Starting VSCode queue processor');
            this.processingInterval = setInterval(() => {
                this.processQueue();
            }, 100); // Check queue every 100ms
        }
    }

    /**
     * Process operation queue
     */
    async processQueue() {
        if (this.isShutdown) {
            return;
        }

        while (this.operationQueue.length > 0 && !this.isShutdown) {
            try {
                const state = this.operationQueue.shift();
                if (!state) {
                    continue;
                }
                await this.processOperation(state);

                // Avoid overly frequent operations
                await new Promise(resolve => setTimeout(resolve, 50));
            } catch (error) {
                this.logger.warn(`Queue processor error occurred:`, error as Error);
            }
        }
    }

    /**
     * Process single operation
     */
    private async processOperation(state: EditorState) {
        try {
            this.sendStateUpdate(state);
        } catch (error) {
            this.logger.warn('Failed to process operation:', error as Error);
        }
    }

    /**
     * Send state update
     */
    private sendStateUpdate(state: EditorState) {
        const messageWrapper = MessageWrapper.create(
            LocalIdentifierManager.getInstance().identifier,
            state
        );

        const success = this.messageSender.sendMessage(messageWrapper);
        if (success) {
            this.logger.info(`✅ Sent message: ${state.action} ${state.filePath}, ${state.getCursorLog()}, ${state.getSelectionLog()}`)
        } else {
            this.logger.info(`❌ Failed to send message: ${state.action} ${state.filePath}, ${state.getCursorLog()}, ${state.getSelectionLog()}`)
        }
    }


    /**
     * Stop queue processor
     */
    dispose() {
        this.logger.info('Starting to shut down VSCode queue processor');

        this.isShutdown = true;

        if (this.processingInterval) {
            clearInterval(this.processingInterval);
            this.processingInterval = null;
        }

        // Clear queue
        this.operationQueue.length = 0;

        this.logger.info('VSCode queue processor has been shut down');
    }
}
